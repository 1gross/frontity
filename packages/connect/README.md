## Description
The state manager of Frontity. Similar to MobX but much more lightweight because it's based on ES6 proxies.

You can find more information about it at [Frontity docs](https://docs.frontity.org/javascript-and-and-react/react-basic#state-manager-frontity-connect).
