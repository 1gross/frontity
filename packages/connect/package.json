{
  "name": "@frontity/connect",
  "version": "0.1.0",
  "description": "A state manager that's easy to use and hackable at the same time.",
  "keywords": [
    "frontity",
    "connect"
  ],
  "homepage": "https://frontity.org",
  "license": "Apache-2.0",
  "repository": {
    "type": "git",
    "url": "https://github.com/frontity/frontity.git"
  },
  "bugs": {
    "url": "https://community.frontity.org"
  },
  "scripts": {
    "test:ci": "jest",
    "test": "jest --watch",
    "test:inspect": "node --inspect ./node_modules/.bin/jest --watch --no-cache --runInBand"
  },
  "dependencies": {
    "recursive-iterator": "^3.3.0"
  },
  "devDependencies": {
    "@babel/core": "^7.4.4",
    "@babel/preset-env": "^7.4.4",
    "@types/jest": "^24.0.11",
    "babel-jest": "^24.8.0",
    "jest": "^24.7.1",
    "ts-jest": "^24.0.2",
    "typescript": "^3.4.2"
  }
}
