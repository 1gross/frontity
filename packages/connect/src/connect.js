import { createContext, useContext, createElement } from "react";
import { view } from "@risingstack/react-easy-state";

/**
 * A React context that stores the store of Frontity Connect.
 */
const context = createContext();

/**
 * The Provider of the store context, to be used in the root of the React
 * application.
 */
export const Provider = context.Provider;

/**
 * Connects a React component with the Frontity connect store.
 *
 * @param Component The React component to be connected
 * @returns The same Component, but with the store included in the props.
 */
export const connect = (Component) => {
  const isStatelessComp = !(
    Component.prototype && Component.prototype.isReactComponent
  );

  if (isStatelessComp) {
    return view((props) => {
      const store = useContext(context);
      return Component({ ...props, ...store });
    });
  } else {
    class ConnectedComponent extends Component {
      static contextType = context;

      render() {
        return createElement(Component, {
          ...this.props,
          ...this.context,
        });
      }
    }

    return view(ConnectedComponent);
  }
};
