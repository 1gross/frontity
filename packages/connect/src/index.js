export { observe, unobserve } from "./observer";
export { observable, isObservable, raw, clone } from "./observable";
export { createStore, getSnapshot } from "./create-store";
export { connect as default, Provider, useConnect } from "./connect";
export { batch } from "./scheduler";
