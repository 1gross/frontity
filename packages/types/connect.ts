import Package from "./package";
import { ResolveState, Omit } from "./utils";

// Resolves the actions for things like Action and Derived.
export type ResolveActions<Actions extends Package["state"]> = {
  [P in keyof Actions]: Actions[P] extends (
    ...a: any
  ) => (arg: infer Arg) => void // Turns "state => args => {}" into "args => {}"
    ? (arg: Arg) => void // Turns "state => {}" into "() => {}"
    : Actions[P] extends (state: Package["state"]) => any
    ? () => void
    : ResolveActions<Actions[P]>
};

export type Connect<S extends Package, Props extends object = {}> = Omit<
  S,
  "state" | "actions" | "name"
> & {
  state: ResolveState<S["state"]>;
  actions: ResolveActions<S["actions"]>;
} & Props;

export default Connect;
