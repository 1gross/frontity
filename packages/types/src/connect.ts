import Package from "./package";
import { ResolveState, Omit, ResolveActions } from "./utils";

export type Connect<S extends Package, Props extends object = {}> = Omit<
  S,
  "state" | "actions" | "name"
> & {
  state: ResolveState<S["state"]>;
  actions: ResolveActions<S["actions"]>;
} & Props;

export type ExternalProps<T extends Package> = Omit<
  T,
  "state" | "actions" | "libraries" | "roots" | "fills"
>;

export default Connect;
