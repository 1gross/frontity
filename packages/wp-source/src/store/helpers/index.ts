export const getIdBySlug = (
  entityMap: {
    [id: number]: {
      slug: string;
      id: number;
    };
  },
  slug: string
): number | undefined => {
  const entity = Object.values(entityMap).find(e => e.slug === slug);
  return entity && entity.id;
};

export const getTotal = (response: Response): number =>
  parseInt(response.headers.get("X-WP-Total"));

export const getTotalPages = (response: Response): number =>
  parseInt(response.headers.get("X-WP-TotalPages"));

export const normalizePath = (pathOrLink: string): string => {
  try {
    return new URL(pathOrLink).pathname;
  } catch (e) {
    // is not a URL
    return pathOrLink;
  }
};
