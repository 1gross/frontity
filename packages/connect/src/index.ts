import { Store, State, ProxifyOptions, Context, Actions } from "../types";

export const proxifyState = <StoreType extends Store>(
  rawStore: StoreType,
  context?: Context,
  options?: ProxifyOptions
): State<StoreType> => {
  return rawStore.state as State<StoreType>; // Overwrite type until it's implemented.
};

export const proxifyActions = <StoreType extends Store>(
  rawStore: StoreType,
  context?: Context,
  options?: ProxifyOptions
): Actions<StoreType> => {
  return rawStore.actions as Actions<StoreType>; // Overwrite type until it's implemented.
};
