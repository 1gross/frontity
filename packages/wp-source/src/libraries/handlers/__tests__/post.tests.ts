import { createStore, InitializedStore } from "@frontity/connect";
import WpSource from "../../../../types";
import handler from "../post";
import { mockResponse } from "./mocks/helpers";

import wpSource from "../../../";
jest.mock("../../../");

const post60 = {
  id: 60,
  date: "2016-11-25T18:31:11",
  slug: "the-beauties-of-gullfoss",
  type: "post",
  link: "https://test.frontity.io/2016/the-beauties-of-gullfoss/",
  author: 4,
  featured_media: 62,
  meta: [],
  categories: [57, 59, 56, 3, 8, 58],
  tags: [10, 9, 13, 11]
};

let store: InitializedStore<WpSource>;

beforeEach(() => {
  const config = wpSource();
  const { api, populate } = config.libraries.source;

  // mock api
  (api.get as jest.Mock).mockResolvedValue(mockResponse([post60]));

  // mock populate
  (populate as jest.Mock).mockResolvedValue([
    {
      id: 60,
      slug: "the-beauties-of-gullfoss",
      link: "https://test.frontity.io/2016/the-beauties-of-gullfoss/"
    }
  ]);

  store = createStore(config);
});

describe("post", () => {
  test("doesn't exist in source.post", async () => {
    // source.fetch("/the-beauties-of-gullfoss/")
    store.state.source.data["/the-beauties-of-gullfoss/"] = {
      isFetching: true,
      isReady: false
    };

    await handler({
      route: "/the-beauties-of-gullfoss/",
      params: { slug: "the-beauties-of-gullfoss" },
      ...store
    });

    expect(store.state.source.data).toMatchSnapshot();
  });

  test("exists in source.post", async () => {
    const get = store.libraries.source.api.get as jest.Mock;

    store.state.source.post[60] = post60;

    // source.fetch("/the-beauties-of-gullfoss/")
    store.state.source.data["/the-beauties-of-gullfoss/"] = {
      isFetching: true,
      isReady: false
    };

    await handler({
      route: "/the-beauties-of-gullfoss/",
      params: { slug: "the-beauties-of-gullfoss" },
      ...store
    });

    expect(get).not.toBeCalled();
    expect(store.state.source.data).toMatchSnapshot();
  });

  test("throws an error if it doesn't exist", async () => {
    store.libraries.source.api.get = jest
      .fn()
      .mockResolvedValueOnce(mockResponse([]));

    store.libraries.source.populate = jest
      .fn()
      .mockResolvedValueOnce(mockResponse([]));

    // source.fetch("/the-beauties-of-gullfoss/")
    store.state.source.data["/the-beauties-of-gullfoss/"] = {
      isFetching: true,
      isReady: false
    };

    const promise = handler({
      route: "/the-beauties-of-gullfoss/",
      params: { slug: "the-beauties-of-gullfoss" },
      ...store
    });

    expect(promise).rejects.toThrowError();
  });

  test("fetchs from a different endpoint with extra params", async () => {
    // change the default post endpoint and params
    store.state.source.postEndpoint = "multiple-post-type";
    store.state.source.params = { type: ["post", "travel"] };

    // source.fetch("/the-beauties-of-gullfoss/")
    store.state.source.data["/the-beauties-of-gullfoss/"] = {
      isFetching: true,
      isReady: false
    };

    await handler({
      route: "/the-beauties-of-gullfoss/",
      params: { slug: "the-beauties-of-gullfoss" },
      ...store
    });

    const apiGet = jest.spyOn(store.libraries.source.api, "get");
    expect(apiGet.mock.calls).toMatchSnapshot();
  });
});
