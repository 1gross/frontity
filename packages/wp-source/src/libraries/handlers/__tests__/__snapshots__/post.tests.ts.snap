// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`post doesn't exist in source.post 1`] = `
Object {
  "/the-beauties-of-gullfoss/": Object {
    "id": 60,
    "isFetching": true,
    "isPost": true,
    "isPostType": true,
    "isReady": false,
    "type": "post",
  },
}
`;

exports[`post exists in source.post 1`] = `
Object {
  "/the-beauties-of-gullfoss/": Object {
    "id": 60,
    "isFetching": true,
    "isPost": true,
    "isPostType": true,
    "isReady": false,
    "type": "post",
  },
}
`;

exports[`post fetchs from a different endpoint with extra params 1`] = `
Array [
  Array [
    Object {
      "endpoint": "multiple-post-type",
      "params": Object {
        "_embed": true,
        "slug": "the-beauties-of-gullfoss",
        "type": Array [
          "post",
          "travel",
        ],
      },
    },
  ],
]
`;
